### PR TITLE
Version 1.2.3

### PR DESCRIPTION
**Pull request by alainpimentel**
- Prevent ConcurrentModificationException with synchronization
- Prevent sending data when there are no active connections
- Add heart beat
- New remote logging to work with bridge monitor

### CODE REVIEW

File: android-wsbridge/app/src/main/java/com/dji/wsbridge/lib/DJILogger.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+public class DJILogger extends Thread {
+
+    public static final String CONNECTION_GOOD = "good";
+    public static final String CONNECTION_BAD = "bad";
+    public static final String CONNTECTION_WARNING = "warning";
+    public static final int CONNECTION_RC = 1;
+    public static final int CONNECTION_BRIDGE = 2;
+    private static final String TAG = "RemoteLogger";
+    // Add your remote server IP
+    private static final String REMOTE_LOGGER_URL = "";
```
Comment (talobin) [inline android-wsbridge/app/src/main/java/com/dji/wsbridge/lib/DJILogger.java]: Should add something like
"your_url_goes_here.com"

File: android-wsbridge/app/build.gradle [hunk 2]
```diff
@@ -36,7 +36,7 @@ android {
 
 dependencies {
     compile 'com.amitshekhar.android:android-networking:0.3.0'
-    compile 'org.java-websocket:java-websocket:1.3.3'
+    compile 'org.java-websocket:java-websocket:1.3.1'
```
Comment (talobin) [inline android-wsbridge/app/build.gradle]: Is it backward change necessary?

Comment (alainpimentel) [reply android-wsbridge/app/build.gradle]: I looked into this and found that in our private repo the library has never been at 1.3.3.
So it was changed in the public repo but not the private one. I'll do some testing to see if it causes any issues and I'll put it back.

File: android-wsbridge/app/build.gradle [hunk 3]
```diff
@@ -36,7 +36,7 @@ android {
 
 dependencies {
     compile 'com.amitshekhar.android:android-networking:0.3.0'
-    compile 'org.java-websocket:java-websocket:1.3.3'
```
Comment (talobin) [inline android-wsbridge/app/build.gradle]: Should probably keep this higher version

File: android-wsbridge/app/src/main/java/com/dji/wsbridge/BridgeActivity.java [hunk 4]
```diff
@@ -285,37 +312,18 @@ private void setupWSConnectionManager() {
...
-            }
-        }
-        if (shouldRefresh) {
-            refresh();
-        }
-        showToast("Network ", isWSConnected.get(), event.getMessage());
-    }
-
-    @Subscribe
-    public void onWSTrafficEvent(WSConnectionManager.WSTrafficEvent event) {
```
Comment (talobin) [inline android-wsbridge/app/src/main/java/com/dji/wsbridge/BridgeActivity.java]: Should add back this logic.

**Review by talobin — CHANGES_REQUESTED** [review]

File: android-wsbridge/app/src/main/java/com/dji/wsbridge/BridgeActivity.java [hunk 5]
```diff
@@ -285,37 +312,18 @@ private void setupWSConnectionManager() {
     public void onWSConnectionEvent(WSConnectionManager.WSConnectionEvent event) {
         boolean shouldRefresh = false;
         if (event.getActiveConnectionCount() > 0) {
-            if (isWSConnected.compareAndSet(false, true)) {
```
Comment (talobin) [inline android-wsbridge/app/src/main/java/com/dji/wsbridge/BridgeActivity.java:288]: Keep this logic

File: android-wsbridge/app/src/main/java/com/dji/wsbridge/lib/connection/WSConnectionManager.java [hunk 6]
```diff
@@ -83,18 +83,13 @@ public void onClose(WebSocket conn, int code, String reason, boolean remote) {
...
-        DJIRemoteLogger.d(TAG, "onError " + ex.getMessage());
+        DJILogger.d(TAG, "onError " + ex.getMessage());
         sendConnectivityStatus(false, conn);
         if (activeConnectionCount() <= 0) {
             stopKeepAlivePolling();
         }
     }
 
-    @Override
-    public void onStart() {
```
Comment (talobin) [inline android-wsbridge/app/src/main/java/com/dji/wsbridge/lib/connection/WSConnectionManager.java]: Keep this

File: android-wsbridge/app/src/main/java/com/dji/wsbridge/lib/connection/WSConnectionManager.java [hunk 7]
```diff
@@ -249,35 +244,24 @@ public ByteBuffer read() {
...
         if (txStatTracker != null) {
             txStatTracker.increaseByteCount(b.length);
         }
-        if (con.size() > 0) {
+        if (activeConnectionCount() > 0) {
             final int maxBufferSizePerConnection = MAX_BUFFER_SIZE / con.size();
-            for (WebSocket conn : con) {
-                if (conn.isOpen()) {
-                    final boolean isSlowTraffic;
-                    if (conn instanceof WebSocketImpl && ((WebSocketImpl) conn).outQueue.size() > maxBufferSizePerConnection) {
```
Comment (talobin) [inline android-wsbridge/app/src/main/java/com/dji/wsbridge/lib/connection/WSConnectionManager.java:262]: Keep this

**Review by talobin — APPROVED** [review]

**Pull request merged**